transformations/close_homophones_swap/transformation.py

```diff
@@ -0,0 +1,59 @@
+import itertools
+import random
+import spacy
+import numpy
+
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+from SoundsLike.SoundsLike import Search
+from spacy.attrs import LOWER, POS, ENT_TYPE, IS_ALPHA
+from spacy.tokens import Doc
+
+"""
+Base Class for implementing the different input transformations a generation should be robust against.
+"""
+
+
+def close_homophones_swap(text, corrupt_prob, seed=0, max_output=1, nlp = None):
+    random.seed(seed)
```

Comment:
I'm not sure but I think random.seed should go inside the for loop so that the seed would work for num_outputs > 1?

Reply:
Hum, I am not sure what the expected behavior should be here? I thought the idea was to have something to set the seed such that the output can be deterministic for the test cases, but we will want something random in the actual generation where num_outputs > 1. If we set the seed every time we generate a new sample, wouldn't reset the random state and output the same thing?

Reply:
oh my bad! What I meant was random.seed works only once so if you are looping for max_output > 1, you might get different results everytime. I think the correct way would be replace random.choice with random.choices below?

Reply:
My understanding of the expected behavior is when max_output > 1, we are supposed to output multiple versions of the augmented sentence from a single source sentence. The current implementation allows it to swap at different locations on the same sentence.

The one you are purposing works, but essentially doing that would make it swap at only a single set of locations for each sentence. so the output could be less diversified.

Reply:
Okay, but why not choose the tokens using random.choices(tokens) then? I'm trying to ensure that the same results turn up for num_outputs > 1 and one seed.

Reply:
I think it does output the same results everytime we run with the same seed, given any max_outputs > 1, since every random function is controlled by random.seed, hence the pesudo random sequence would be the same.

To choose tokens to be swapped with random.choices will definitely look more elegant though. But since we are calling the homophones similarity function on each token, it would be more straightforward to implement it with a for loop, which loops through each token.

Reply:
I just ran test with max_output = 3 for multiple times, it looks like it does output same set of results everytime.

Reply:
I had the same question a while ago and it turns out that you only need to call `random.seed()` once. The next `random()` calls return different numbers but it will always be the same sequence of numbers for that seed.